### PR TITLE
[WIP] Attempted conversion of i18n package

### PR DESCRIPTION
**Pull request by barklund**
## Context

This is a first attempt at converting the i18n package to TypeScript.

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #

### CODE REVIEW

File: packages/i18n/src/translateWithMarkup.tsx [hunk 1]
```diff
@@ -52,18 +58,22 @@ const VOID_ELEMENTS = [
...
+  mapping: Mapping = {}
+): React.ReactNode {
+  const { firstChild, nodeType, textContent } = node;
   if (Node.TEXT_NODE === nodeType) {
     return textContent;
   }
 
-  const children = node.hasChildNodes()
-    ? [...childNodes].map((child) => transform(child, mapping))
-    : null;
```
Comment (barklund) [inline packages/i18n/src/translateWithMarkup.tsx:63]: I am pretty sure this old code is wrong, no? It loops over all the child nodes and transforms each, but `transformNode` transform both the given node and all its following siblings, so this old code duplicates a lot of nodes, no?

Well, at least the type system told me that the `children` variable had type `ReactNode[][]`, which definitely seemed wrong.

File: packages/i18n/src/translateWithMarkup.tsx [hunk 2]
```diff
@@ -52,18 +58,22 @@ const VOID_ELEMENTS = [
...
+  const children = node.hasChildNodes() ? transform(firstChild, mapping) : null;
 
-  if (localName in mapping) {
-    return cloneElement(mapping[localName], null, children);
+  if ('localName' in node) {
+    const { localName } = node as Element;
+    if (localName in mapping) {
+      return cloneElement(mapping[localName], undefined, children);
```
Comment (barklund) [inline packages/i18n/src/translateWithMarkup.tsx:75]: For the life of me, I can't figure out how to get TS to accept the type of this one? I tried everything, but I can't match the types to [any of the overloads available](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L292-L324). And similarly for `createElement` and [its overloads](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L254-L288) a few lines below.

**Comment by swissspidy** [comment]
@barklund Did you miss #12198?

**Comment by barklund** [comment]
> @barklund Did you miss #12198?

Obviously! Sorry about that - I did try to search for it, but clearly weren't that good at it. 🤦 

Luckily I didn't spend too much time on this, and maybe some of the work can be reused? I'll check your PR later and close this one.

**Comment by swissspidy** [comment]
Merged #12198 now. Feel free to rebase your PR if you have any suggestions for improvements!

**Pull request closed without merging**